process/block/preprocess/rewardsHandler.go

```diff
@@ -8,79 +8,182 @@ import (
 		return nil, process.ErrNilHasher
 	}
 	if marshalizer == nil || marshalizer.IsInterfaceNil() {
 		return nil, process.ErrNilMarshalizer
 	}
+	if shardCoordinator == nil || shardCoordinator.IsInterfaceNil() {
+		return nil, process.ErrNilShardCoordinator
+	}
+	if adrConv == nil || adrConv.IsInterfaceNil() {
+		return nil, process.ErrNilAddressConverter
+	}
+	if store == nil || store.IsInterfaceNil() {
+		return nil, process.ErrNilStorage
+	}
+	if rewardTxPool == nil || rewardTxPool.IsInterfaceNil() {
+		return nil, process.ErrNilRewardTxDataPool
+	}
 
 	rtxh := &rewardsHandler{
 		address:          address,
 		shardCoordinator: shardCoordinator,
+		adrConv:          adrConv,
 		hasher:           hasher,
 		marshalizer:      marshalizer,
+		store:            store,
+		rewardTxPool:     rewardTxPool,
 	}
+
 	rtxh.accumulatedFees = big.NewInt(0)
-	rtxh.rewardTxsFromBlock = make(map[string]*rewardTx.RewardTx)
+	rtxh.rewardTxsForBlock = make(map[string]*rewardTx.RewardTx)
 
 	return rtxh, nil
 }
 
 // SaveCurrentIntermediateTxToStorage saves current cached data into storage - already saved for txs
 func (rtxh *rewardsHandler) SaveCurrentIntermediateTxToStorage() error {
-	//TODO implement me - save only created accumulatedFees
+	rtxh.mut.Lock()
+	defer rtxh.mut.Unlock()
+
+	for _, rTx := range rtxh.rewardTxsForBlock {
+		buff, err := rtxh.marshalizer.Marshal(rTx)
+		if err != nil {
+			return err
+		}
+
+		errNotCritical := rtxh.store.Put(dataRetriever.RewardTransactionUnit, rtxh.hasher.Compute(string(buff)), buff)
+		if errNotCritical != nil {
+			log.Error(errNotCritical.Error())
+		}
+	}
+
 	return nil
 }
 
 // AddIntermediateTransactions adds intermediate transactions to local cache
 func (rtxh *rewardsHandler) AddIntermediateTransactions(txs []data.TransactionHandler) error {
+	rtxh.mut.Lock()
+	defer rtxh.mut.Unlock()
+
+	for i := 0; i < len(txs); i++ {
+		addedRewardTx, ok := txs[i].(*rewardTx.RewardTx)
+		if !ok {
+			return process.ErrWrongTypeAssertion
+		}
+
+		if addedRewardTx.ShardId != rtxh.shardCoordinator.SelfId() {
+			continue
+		}
+
+		rewardTxHash, err := core.CalculateHash(rtxh.marshalizer, rtxh.hasher, txs[i])
+		if err != nil {
+			return err
+		}
+
+		rtxh.rewardTxsForBlock[string(rewardTxHash)] = addedRewardTx
+	}
+
 	return nil
 }
 
+func (rtxh *rewardsHandler) getShardIdsFromAddress(addr []byte) (uint32, error) {
+	address, err := rtxh.adrConv.CreateAddressFromPublicKeyBytes(addr)
+	if err != nil {
+		return rtxh.shardCoordinator.NumberOfShards(), err
+	}
+	shardId := rtxh.shardCoordinator.ComputeId(address)
+
+	return shardId, nil
+}
+
 // CreateAllInterMiniBlocks creates miniblocks from process transactions
 func (rtxh *rewardsHandler) CreateAllInterMiniBlocks() map[uint32]*block.MiniBlock {
-	calculatedRewardTxs := rtxh.CreateAllUTxs()
+	rtxh.mutGenRewardTxs.Lock()
+	calculatedRewardTxs := make([]data.TransactionHandler, 0)
+	rtxh.feeRewards = rtxh.createRewardFromFees()
+	rtxh.addTransactionsToPool(rtxh.feeRewards)
```

Comment:
if we go with the SCR way of processing there is not need to add the created transactions into pool.